{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "N째 of unique orthogroups: 9\n",
      "N째 of colors 16\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "import seaborn as sns\n",
    "\n",
    "\n",
    "map = {'../results/PipCYP719A37/hits09.pep': '../data/proteomes/piper09/proteome09.csv',\n",
    "       '../results/PipCYP719A37/hits23.pep': '../data/proteomes/piper23/proteome23.csv',\n",
    "       '../results/PipCYP719A37/hits25.pep': '../data/proteomes/piper25/proteome25.csv',\n",
    "       '../results/PipCYP719A37/hits26.pep': '../data/proteomes/piper26/proteome26.csv',\n",
    "       '../results/PipCYP719A37/hits27.pep': '../data/proteomes/piper27/proteome27.csv'}\n",
    "\n",
    "for hits_path, proteome_path in map.items():\n",
    "\n",
    "       #import BLAST hits\n",
    "       hits = pd.Series([record.description for record in SeqIO.parse(hits_path, 'fasta')]).to_frame(name='full_id')\n",
    "       hits.set_index(hits['full_id'].apply(lambda x: x.split('_cov')[0]), inplace=True)\n",
    "\n",
    "       #import proteome\n",
    "       proteome = pd.read_csv(proteome_path, index_col='protein_id')\n",
    "       proteome['mean'] = proteome.loc[:, [col for col in proteome.columns if 'rep' in col]].mean(axis=1).round(decimals=0) #calculate mean\n",
    "       proteome['stdev'] = proteome.loc[:, [col for col in proteome.columns if 'rep' in col]].std(axis=1).round(decimals=0) #calculate stdev\n",
    "          \n",
    "       #add mean column to hits header\n",
    "       if '_expr' not in hits['full_id'].iloc[0]:\n",
    "              mean = proteome.loc[[id for id in hits.index if id in proteome.index], 'mean']\n",
    "              hits = pd.merge(hits, mean, how='left', left_index=True, right_index=True)\n",
    "              hits['full_id'] = hits['full_id'] + '_expr' + hits['mean'].astype(int).astype(str)\n",
    "\n",
    "\n",
    "       # Update headers using SeqIO\n",
    "       records = list(SeqIO.parse(hits_path, 'fasta'))\n",
    "       \n",
    "       for i, record in enumerate(records):\n",
    "              record.id = hits['full_id'][i]\n",
    "              record.description = ''\n",
    "              record.name = ''        \n",
    "              \n",
    "       SeqIO.write(records, hits_path, 'fasta-2line')\n",
    "\n",
    "\n",
    "\n",
    "# add mean column to hits header if not already present\n",
    "if '_expr' not in hits['full_id'].iloc[0]:\n",
    "     mean = proteome.loc[[id for id in hits.index if id in proteome.index], 'mean']\n",
    "     hits = pd.merge(hits, mean, how='left', left_index=True, right_index=True)\n",
    "     hits['full_id'] = hits['full_id'] + '_expr' + hits['mean'].astype(int).astype(str)\n",
    "\n",
    "\n",
    "#concatenate hits\n",
    "all_hits =pd.DataFrame()\n",
    "\n",
    "for hits_path in map.keys():\n",
    "    hits = pd.Series([record.id for record in SeqIO.parse(hits_path, 'fasta')]).to_frame(name='full_id')\n",
    "    hits['short_id'] = hits['full_id'].apply(lambda x: x.split('_cov')[0]) #short id\n",
    "\n",
    "    all_hits = pd.concat([all_hits, hits], axis=0).reset_index(drop=True)\n",
    "\n",
    "all_hits['orthog_id'] = '' #add orthogroup_id column\n",
    "\n",
    "#import sonicparanoid orthogroups\n",
    "sonicparanoid_path = '../results/orthologs/runs/piperNET/ortholog_groups/ortholog_groups.tsv'\n",
    "sonicparanoid = pd.read_csv(sonicparanoid_path, sep='\\t')\n",
    "\n",
    "\n",
    "#extract orthogourp id for each hit\n",
    "legend = {'Pfim': 'proteome09.pep',\n",
    "          'Parb': 'proteome23.pep',\n",
    "          'Pama': 'proteome25.pep',\n",
    "          'Psar': 'proteome26.pep',\n",
    "          'Pkad': 'proteome27.pep'}\n",
    "\n",
    "for species, proteome in legend.items():\n",
    "\n",
    "    for header in all_hits['short_id']:\n",
    "        \n",
    "        if species in header:\n",
    "            orthog_id = sonicparanoid.loc[sonicparanoid[proteome].str.contains(header), ['group_id']] #single-entry dataframe\n",
    "            if not orthog_id.empty:\n",
    "                all_hits.loc[all_hits['short_id'] == header, 'orthog_id'] = int(orthog_id.iloc[0])\n",
    "\n",
    "\n",
    "#Export all_hits dataframe\n",
    "all_hits.to_csv('../results/PipCYP719A37/orthogroups.csv', index=False)\n",
    "\n",
    "\n",
    "###Create itol mapping file\n",
    "itol = all_hits.copy()\n",
    "\n",
    "#assign colors to orthogroups\n",
    "unique_orthog = itol['orthog_id'].unique()\n",
    "palette = ['#2ECC71', '#DAF7A6', '#B7950B', '#FFACAC', '#FF0000', '#48C9B0', '#FBFF00', '#6C3483', '#797D7F', '#85C1E9', '#F7DC6F', '#AEB6BF', '#0E6251', '#E67E22', '#3333FF', '#FF2DF5']\n",
    "colors = sns.color_palette(palette, n_colors=len(unique_orthog))\n",
    "color_map = dict(zip(unique_orthog, colors)) #create dictionary\n",
    "itol['color'] = itol['orthog_id'].map(color_map)\n",
    "itol['color'] = itol['color'].apply(lambda x: f'#{int(x[0]*255):02x}{int(x[1]*255):02x}{int(x[2]*255):02x}') #convert to hex format\n",
    "itol.loc[itol['orthog_id']  == '', 'color'] = '#ffffff' #set to white hits with no orthog_id\n",
    "\n",
    "#adjust to itol template\n",
    "itol_header = 'TREE_COLORS\\nSEPARATOR TAB\\nDATA\\n'\n",
    "itol_text = pd.concat([itol['full_id'], pd.Series(['range'] * len(itol), name='type'), itol['color'], itol['orthog_id'], pd.Series(['orthogroup'] * len(itol), name='label')], axis=1)\n",
    "itol_file = itol_header + itol_text.to_csv(sep='\\t', header=False, index=False)\n",
    "\n",
    "#write file\n",
    "with open('../results/PipCYP719A37/itol.txt', 'w') as file:\n",
    "    file.write(itol_file)\n",
    "\n",
    "#log\n",
    "print('N째 of unique orthogroups:', len(unique_orthog))\n",
    "print('N째 of colors', len(palette))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pNET_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  },
  "metadata": {
   "interpreter": {
    "hash": "ac59ebe37160ed0dfa835113d9b8498d9f09ceb179beaac4002f036b9467c963"
   }
  },
  "orig_nbformat": 2
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
