{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#set cwd to /PiperNET\n",
    "import os\n",
    "os.chdir('..')\n",
    "\n",
    "import subprocess\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import argparse\n",
    "from Bio import SeqIO\n",
    "from io import StringIO\n",
    "from src.rnaseq_utils import get_config\n",
    "from src.rnaseq_utils import download_enzymeDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "'enzyme_seq' folder already exists at data/rna-seq/enzyme_seq\n",
      "AtPAL2 sequence already present at 'data/rna-seq/enzyme_seq/AtPAL2.fasta'\n",
      "PaPAL sequence already present at 'data/rna-seq/enzyme_seq/PaPAL.fasta'\n",
      "AtC4H sequence already present at 'data/rna-seq/enzyme_seq/AtC4H.fasta'\n",
      "MsC4H sequence already present at 'data/rna-seq/enzyme_seq/MsC4H.fasta'\n",
      "At4CL1 sequence already present at 'data/rna-seq/enzyme_seq/At4CL1.fasta'\n",
      "At4CL2 sequence already present at 'data/rna-seq/enzyme_seq/At4CL2.fasta'\n",
      "At4CL3 sequence already present at 'data/rna-seq/enzyme_seq/At4CL3.fasta'\n",
      "At4CL4 sequence already present at 'data/rna-seq/enzyme_seq/At4CL4.fasta'\n",
      "Nt4CL1 sequence already present at 'data/rna-seq/enzyme_seq/Nt4CL1.fasta'\n",
      "Nt4CL2 sequence already present at 'data/rna-seq/enzyme_seq/Nt4CL2.fasta'\n",
      "HlCL1 sequence already present at 'data/rna-seq/enzyme_seq/HlCL1.fasta'\n",
      "PhCL1 sequence already present at 'data/rna-seq/enzyme_seq/PhCL1.fasta'\n",
      "NpCL1 sequence already present at 'data/rna-seq/enzyme_seq/NpCL1.fasta'\n",
      "NpCL2 sequence already present at 'data/rna-seq/enzyme_seq/NpCL2.fasta'\n",
      "PcCL2 sequence already present at 'data/rna-seq/enzyme_seq/PcCL2.fasta'\n",
      "PcCL2 sequence already present at 'data/rna-seq/enzyme_seq/PcCL2.fasta'\n",
      "PtCL sequence already present at 'data/rna-seq/enzyme_seq/PtCL.fasta'\n",
      "At4CLL1 sequence already present at 'data/rna-seq/enzyme_seq/At4CLL1.fasta'\n",
      "At4CLL2 sequence already present at 'data/rna-seq/enzyme_seq/At4CLL2.fasta'\n",
      "At4CLL3 sequence already present at 'data/rna-seq/enzyme_seq/At4CLL3.fasta'\n",
      "At4CLL6 sequence already present at 'data/rna-seq/enzyme_seq/At4CLL6.fasta'\n",
      "At4CLL7 sequence already present at 'data/rna-seq/enzyme_seq/At4CLL7.fasta'\n",
      "At4CLL8 sequence already present at 'data/rna-seq/enzyme_seq/At4CLL8.fasta'\n",
      "AtCCOAOMT1 sequence already present at 'data/rna-seq/enzyme_seq/AtCCOAOMT1.fasta'\n",
      "NtCCOAOMT1 sequence already present at 'data/rna-seq/enzyme_seq/NtCCOAOMT1.fasta'\n",
      "PnCYP719A37 sequence already present at 'data/rna-seq/enzyme_seq/PnCYP719A37.fasta'\n",
      "PnPIPCoAL sequence already present at 'data/rna-seq/enzyme_seq/PnPIPCoAL.fasta'\n",
      "PnBAHD1 sequence already present at 'data/rna-seq/enzyme_seq/PnBAHD1.fasta'\n",
      "PnBAHD2 sequence already present at 'data/rna-seq/enzyme_seq/PnBAHD2.fasta'\n",
      "PmCHS sequence already present at 'data/rna-seq/enzyme_seq/PmCHS.fasta'\n",
      "AtCHS sequence already present at 'data/rna-seq/enzyme_seq/AtCHS.fasta'\n",
      "PaCHS sequence already present at 'data/rna-seq/enzyme_seq/PaCHS.fasta'\n",
      "AtPKS1 sequence already present at 'data/rna-seq/enzyme_seq/AtPKS1.fasta'\n",
      "AtPKS2 sequence already present at 'data/rna-seq/enzyme_seq/AtPKS2.fasta'\n",
      "AtPKS3 sequence already present at 'data/rna-seq/enzyme_seq/AtPKS3.fasta'\n",
      "PmSPS1 sequence already present at 'data/rna-seq/enzyme_seq/PmSPS1.fasta'\n",
      "PmSPS2 sequence already present at 'data/rna-seq/enzyme_seq/PmSPS2.fasta'\n",
      "VvSTS1 sequence already present at 'data/rna-seq/enzyme_seq/VvSTS1.fasta'\n",
      "AhSTS1 sequence already present at 'data/rna-seq/enzyme_seq/AhSTS1.fasta'\n",
      "ClCURS1 sequence already present at 'data/rna-seq/enzyme_seq/ClCURS1.fasta'\n",
      "ClCURS2 sequence already present at 'data/rna-seq/enzyme_seq/ClCURS2.fasta'\n",
      "ClCURS3 sequence already present at 'data/rna-seq/enzyme_seq/ClCURS3.fasta'\n",
      "ClDCS sequence already present at 'data/rna-seq/enzyme_seq/ClDCS.fasta'\n",
      "OsCUR sequence already present at 'data/rna-seq/enzyme_seq/OsCUR.fasta'\n",
      "AaPCS sequence already present at 'data/rna-seq/enzyme_seq/AaPCS.fasta'\n",
      "AaOKS1 sequence already present at 'data/rna-seq/enzyme_seq/AaOKS1.fasta'\n",
      "AaOKS2 sequence already present at 'data/rna-seq/enzyme_seq/AaOKS2.fasta'\n",
      "AaOKS3 sequence already present at 'data/rna-seq/enzyme_seq/AaOKS3.fasta'\n",
      "AaPKS sequence already present at 'data/rna-seq/enzyme_seq/AaPKS.fasta'\n",
      "OsARAS1 sequence already present at 'data/rna-seq/enzyme_seq/OsARAS1.fasta'\n",
      "OsARAS2 sequence already present at 'data/rna-seq/enzyme_seq/OsARAS2.fasta'\n"
     ]
    }
   ],
   "source": [
    "#download enzymeDB\n",
    "enzymeDB_path = 'data/rna-seq/enzymeDB.csv'\n",
    "enzyme_seq_path = 'data/rna-seq/enzyme_seq'\n",
    "download_enzymeDB(enzymeDB_path, enzyme_seq_path)\n",
    "\n",
    "\n",
    "#load sample folder paths from config.yaml\n",
    "config_path = Path('config.yaml')\n",
    "blastDB_paths = get_config(config_path, data='blast', db_dirs='blastDB')\n",
    "proteome_paths = get_config(config_path, data='rna-seq', format='proteomes.csv')\n",
    "\n",
    "\n",
    "#import proteomes\n",
    "proteome_list = {}\n",
    "for sample, path in proteome_paths.items():\n",
    "    try:\n",
    "        proteome_list[sample] = pd.read_csv(path, index_col='id')\n",
    "    except FileNotFoundError:\n",
    "        print(f'{path} not found for sample {sample}. Skipping to the next file.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "########## BLASTp search ##########\n",
    "\n",
    "def run_blastp(seq_paths, blastDB_paths, min_cov = 0, min_sim = 0):\n",
    "    \n",
    "    '''To document'''\n",
    "    \n",
    "    hits = pd.DataFrame()  # empty DataFrame\n",
    "    \n",
    "    for query in seq_paths.values():\n",
    "        for sample, blastDB in zip(blastDB_paths.keys(), blastDB_paths.values()):\n",
    "            print(f'Querying {query} in {sample} transcriptome...')\n",
    "\n",
    "            #construct command\n",
    "            command = f'blastp -db {blastDB} -query {query} -evalue 1e-50 -outfmt \"10 delim=, qacc sseqid qcovs score evalue pident ppos sseq\"'\n",
    "\n",
    "            #run command\n",
    "            output = subprocess.run(command, shell=True, capture_output=True, text=True)\n",
    "\n",
    "            #convert to dataframe\n",
    "            if output.returncode == 0:\n",
    "                output = output.stdout\n",
    "                output = pd.read_csv(StringIO(output), header=None, names=['query', 'hit', 'coverage', 'score', 'evalue', 'identity', 'similarity', 'sequence'])\n",
    "\n",
    "                #create full_id column\n",
    "                output['blast_id'] = output.apply(lambda row: f\"{row['hit']}_cov{row['coverage']}_sim{round(row['similarity'])}\", axis=1)\n",
    "                \n",
    "                #print number of retrieved hits\n",
    "                print(f\"{output.shape[0]} hits found for {query.stem} in {sample} transcriptome.\")\n",
    "\n",
    "                #filter by coverage and similarity\n",
    "                filt = (output['coverage'] > min_cov) & (output['similarity'] > min_sim)\n",
    "                output = output.loc[filt]\n",
    "\n",
    "                #append output to blast_results\n",
    "                hits = pd.concat([hits, output], ignore_index=True)\n",
    "\n",
    "            else:\n",
    "                print(f\"An error occurred while running the BLASTp command!\")\n",
    "\n",
    "    #remove duplicates\n",
    "    hits['query'] = hits['query'].apply(lambda x: [x]) #convert to list\n",
    "    grouped = hits.groupby('hit').agg({'query': 'sum'})\n",
    "    hits = hits.drop(columns='query')\n",
    "    hits = pd.merge(hits, grouped, left_on='hit', right_index=True)\n",
    "    hits['query'] = hits['query'].apply(set) #convert to set\n",
    "\n",
    "    return hits"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Querying data/rna-seq/enzyme_seq/PmSPS1.fasta in piper09 transcriptome...\n",
      "14 hits found for PmSPS1 in piper09 transcriptome.\n",
      "Querying data/rna-seq/enzyme_seq/PmSPS1.fasta in piper12 transcriptome...\n",
      "8 hits found for PmSPS1 in piper12 transcriptome.\n",
      "Querying data/rna-seq/enzyme_seq/PmSPS1.fasta in piper56 transcriptome...\n",
      "14 hits found for PmSPS1 in piper56 transcriptome.\n",
      "Querying data/rna-seq/enzyme_seq/VvSTS1.fasta in piper09 transcriptome...\n",
      "14 hits found for VvSTS1 in piper09 transcriptome.\n",
      "Querying data/rna-seq/enzyme_seq/VvSTS1.fasta in piper12 transcriptome...\n",
      "7 hits found for VvSTS1 in piper12 transcriptome.\n",
      "Querying data/rna-seq/enzyme_seq/VvSTS1.fasta in piper56 transcriptome...\n",
      "14 hits found for VvSTS1 in piper56 transcriptome.\n",
      "Querying data/rna-seq/enzyme_seq/ClCURS1.fasta in piper09 transcriptome...\n",
      "13 hits found for ClCURS1 in piper09 transcriptome.\n",
      "Querying data/rna-seq/enzyme_seq/ClCURS1.fasta in piper12 transcriptome...\n",
      "7 hits found for ClCURS1 in piper12 transcriptome.\n",
      "Querying data/rna-seq/enzyme_seq/ClCURS1.fasta in piper56 transcriptome...\n",
      "13 hits found for ClCURS1 in piper56 transcriptome.\n"
     ]
    }
   ],
   "source": [
    "#PKS\n",
    "enzyme_class = 'PKS'\n",
    "seq_paths = get_config(config_path, data='blast', enzymeDB='enzyme_seq', enzyme_class=enzyme_class)\n",
    "pks = run_blastp(seq_paths, blastDB_paths, min_cov=50, min_sim=50)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>hit</th>\n",
       "      <th>coverage</th>\n",
       "      <th>score</th>\n",
       "      <th>evalue</th>\n",
       "      <th>identity</th>\n",
       "      <th>similarity</th>\n",
       "      <th>sequence</th>\n",
       "      <th>blast_id</th>\n",
       "      <th>query</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Pfim_g13608_i1.p1</td>\n",
       "      <td>99</td>\n",
       "      <td>1763</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>83.038</td>\n",
       "      <td>90.38</td>\n",
       "      <td>MSKTVEEIQAAQRARGPATVLAIGTATPANVVYQADYPDYYFRITK...</td>\n",
       "      <td>Pfim_g13608_i1.p1_cov99_sim90</td>\n",
       "      <td>{PmSPS1, VvSTS1, ClCURS1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td>Pfim_g13608_i1.p1</td>\n",
       "      <td>98</td>\n",
       "      <td>1615</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>74.870</td>\n",
       "      <td>88.86</td>\n",
       "      <td>TVEEIQAAQRARGPATVLAIGTATPANVVYQADYPDYYFRITKSEH...</td>\n",
       "      <td>Pfim_g13608_i1.p1_cov98_sim89</td>\n",
       "      <td>{PmSPS1, VvSTS1, ClCURS1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>Pfim_g13608_i1.p1</td>\n",
       "      <td>98</td>\n",
       "      <td>1282</td>\n",
       "      <td>8.710000e-177</td>\n",
       "      <td>57.180</td>\n",
       "      <td>77.55</td>\n",
       "      <td>IQAAQRARGPATVLAIGTATPANVVYQADYPDYYFRITKSEHMTEL...</td>\n",
       "      <td>Pfim_g13608_i1.p1_cov98_sim78</td>\n",
       "      <td>{PmSPS1, VvSTS1, ClCURS1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Pfim_g19684_i0.p1</td>\n",
       "      <td>97</td>\n",
       "      <td>1686</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>80.879</td>\n",
       "      <td>89.66</td>\n",
       "      <td>TVDDIRKAQRAEGPATVLAIGTATPANCVYQDDYPDYYFRITNSEH...</td>\n",
       "      <td>Pfim_g19684_i0.p1_cov97_sim90</td>\n",
       "      <td>{PmSPS1, VvSTS1, ClCURS1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>Pfim_g19684_i0.p1</td>\n",
       "      <td>99</td>\n",
       "      <td>1641</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>76.093</td>\n",
       "      <td>88.17</td>\n",
       "      <td>MATVDDIRKAQRAEGPATVLAIGTATPANCVYQDDYPDYYFRITNS...</td>\n",
       "      <td>Pfim_g19684_i0.p1_cov99_sim88</td>\n",
       "      <td>{PmSPS1, VvSTS1, ClCURS1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>Pnig_g20709_i0.p1</td>\n",
       "      <td>52</td>\n",
       "      <td>794</td>\n",
       "      <td>5.740000e-105</td>\n",
       "      <td>74.146</td>\n",
       "      <td>81.95</td>\n",
       "      <td>NRYFHQTEELLEKNLNICAYMAPSLDARQDIAVEEVPKLAKEAATT...</td>\n",
       "      <td>Pnig_g20709_i0.p1_cov52_sim82</td>\n",
       "      <td>{PmSPS1, VvSTS1, ClCURS1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>Pnig_g20709_i0.p1</td>\n",
       "      <td>51</td>\n",
       "      <td>734</td>\n",
       "      <td>4.510000e-96</td>\n",
       "      <td>67.000</td>\n",
       "      <td>82.00</td>\n",
       "      <td>RYFHQTEELLEKNLNICAYMAPSLDARQDIAVEEVPKLAKEAATTA...</td>\n",
       "      <td>Pnig_g20709_i0.p1_cov51_sim82</td>\n",
       "      <td>{PmSPS1, VvSTS1, ClCURS1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>Pnig_g20709_i0.p1</td>\n",
       "      <td>52</td>\n",
       "      <td>592</td>\n",
       "      <td>1.320000e-74</td>\n",
       "      <td>52.217</td>\n",
       "      <td>70.44</td>\n",
       "      <td>NRYFHQTEELLEKNLNICAYMAPSLDARQDIAVEEVPKLAKEAATT...</td>\n",
       "      <td>Pnig_g20709_i0.p1_cov52_sim70</td>\n",
       "      <td>{PmSPS1, VvSTS1, ClCURS1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>Pnig_g44643_i0.p1</td>\n",
       "      <td>56</td>\n",
       "      <td>449</td>\n",
       "      <td>2.660000e-53</td>\n",
       "      <td>42.411</td>\n",
       "      <td>61.61</td>\n",
       "      <td>LCKTTTVKTRYVVMSEEILNKYPELAVEGLPTVKQRLEVCNDAVTQ...</td>\n",
       "      <td>Pnig_g44643_i0.p1_cov56_sim62</td>\n",
       "      <td>{PmSPS1, VvSTS1}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>Pnig_g44643_i0.p1</td>\n",
       "      <td>57</td>\n",
       "      <td>471</td>\n",
       "      <td>1.380000e-56</td>\n",
       "      <td>40.807</td>\n",
       "      <td>62.78</td>\n",
       "      <td>LCKTTTVKTRYVVMSEEILNKYPELAVEGLPTVKQRLEVCNDAVTQ...</td>\n",
       "      <td>Pnig_g44643_i0.p1_cov57_sim63</td>\n",
       "      <td>{PmSPS1, VvSTS1}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>82 rows Ã— 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                  hit  coverage  score         evalue  identity  similarity  \\\n",
       "0   Pfim_g13608_i1.p1        99   1763   0.000000e+00    83.038       90.38   \n",
       "31  Pfim_g13608_i1.p1        98   1615   0.000000e+00    74.870       88.86   \n",
       "59  Pfim_g13608_i1.p1        98   1282  8.710000e-177    57.180       77.55   \n",
       "1   Pfim_g19684_i0.p1        97   1686   0.000000e+00    80.879       89.66   \n",
       "30  Pfim_g19684_i0.p1        99   1641   0.000000e+00    76.093       88.17   \n",
       "..                ...       ...    ...            ...       ...         ...   \n",
       "27  Pnig_g20709_i0.p1        52    794  5.740000e-105    74.146       81.95   \n",
       "54  Pnig_g20709_i0.p1        51    734   4.510000e-96    67.000       82.00   \n",
       "81  Pnig_g20709_i0.p1        52    592   1.320000e-74    52.217       70.44   \n",
       "28  Pnig_g44643_i0.p1        56    449   2.660000e-53    42.411       61.61   \n",
       "55  Pnig_g44643_i0.p1        57    471   1.380000e-56    40.807       62.78   \n",
       "\n",
       "                                             sequence  \\\n",
       "0   MSKTVEEIQAAQRARGPATVLAIGTATPANVVYQADYPDYYFRITK...   \n",
       "31  TVEEIQAAQRARGPATVLAIGTATPANVVYQADYPDYYFRITKSEH...   \n",
       "59  IQAAQRARGPATVLAIGTATPANVVYQADYPDYYFRITKSEHMTEL...   \n",
       "1   TVDDIRKAQRAEGPATVLAIGTATPANCVYQDDYPDYYFRITNSEH...   \n",
       "30  MATVDDIRKAQRAEGPATVLAIGTATPANCVYQDDYPDYYFRITNS...   \n",
       "..                                                ...   \n",
       "27  NRYFHQTEELLEKNLNICAYMAPSLDARQDIAVEEVPKLAKEAATT...   \n",
       "54  RYFHQTEELLEKNLNICAYMAPSLDARQDIAVEEVPKLAKEAATTA...   \n",
       "81  NRYFHQTEELLEKNLNICAYMAPSLDARQDIAVEEVPKLAKEAATT...   \n",
       "28  LCKTTTVKTRYVVMSEEILNKYPELAVEGLPTVKQRLEVCNDAVTQ...   \n",
       "55  LCKTTTVKTRYVVMSEEILNKYPELAVEGLPTVKQRLEVCNDAVTQ...   \n",
       "\n",
       "                         blast_id                      query  \n",
       "0   Pfim_g13608_i1.p1_cov99_sim90  {PmSPS1, VvSTS1, ClCURS1}  \n",
       "31  Pfim_g13608_i1.p1_cov98_sim89  {PmSPS1, VvSTS1, ClCURS1}  \n",
       "59  Pfim_g13608_i1.p1_cov98_sim78  {PmSPS1, VvSTS1, ClCURS1}  \n",
       "1   Pfim_g19684_i0.p1_cov97_sim90  {PmSPS1, VvSTS1, ClCURS1}  \n",
       "30  Pfim_g19684_i0.p1_cov99_sim88  {PmSPS1, VvSTS1, ClCURS1}  \n",
       "..                            ...                        ...  \n",
       "27  Pnig_g20709_i0.p1_cov52_sim82  {PmSPS1, VvSTS1, ClCURS1}  \n",
       "54  Pnig_g20709_i0.p1_cov51_sim82  {PmSPS1, VvSTS1, ClCURS1}  \n",
       "81  Pnig_g20709_i0.p1_cov52_sim70  {PmSPS1, VvSTS1, ClCURS1}  \n",
       "28  Pnig_g44643_i0.p1_cov56_sim62           {PmSPS1, VvSTS1}  \n",
       "55  Pnig_g44643_i0.p1_cov57_sim63           {PmSPS1, VvSTS1}  \n",
       "\n",
       "[82 rows x 9 columns]"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#retrieve hits from transcriptomes\n",
    "data = pd.DataFrame()  # empty DataFrame\n",
    "\n",
    "for sample, proteome in zip(proteome_list.keys(), proteome_list.values()):\n",
    "    hits = proteome.loc[proteome.index.isin(pks['hit'])]\n",
    "\n",
    "    #append output to blast_results\n",
    "    data = pd.concat([data, hits])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "sequence               MLVHGSLFRCTPKDHGRGGDQAGHGSRREGHQGMGAARIQDYPPRL...\n",
       "length                                                               310\n",
       "orf_type                                                        complete\n",
       "orientation                                                          (-)\n",
       "orf_score                                                          35.89\n",
       "transcript_length                                                   1333\n",
       "transcript_coverage                                            11.983333\n",
       "pfam                                                          PF02797.19\n",
       "domain                 Chalcone and stilbene synthases, C-terminal do...\n",
       "pfam_e-value                                                         0.0\n",
       "blastp                 sp|A0A2Z5QL08|CHS_RHODA Chalcone synthase OS=R...\n",
       "blastp_e-value                                                       0.0\n",
       "signalp                                                            OTHER\n",
       "targetp                                                             noTP\n",
       "tmhmm                                                                  0\n",
       "topology                                                               o\n",
       "P09_leaf_rep1                                                      0.236\n",
       "P09_leaf_rep2                                                      0.785\n",
       "P09_leaf_rep3                                                      0.319\n",
       "P09_root_rep1                                                      1.514\n",
       "P09_root_rep2                                                      0.626\n",
       "P09_root_rep3                                                      1.453\n",
       "P09_stem_rep1                                                      1.322\n",
       "P09_stem_rep2                                                       1.48\n",
       "P09_stem_rep3                                                      0.689\n",
       "P12_leaf_rep1                                                        NaN\n",
       "P12_leaf_rep2                                                        NaN\n",
       "P12_leaf_rep3                                                        NaN\n",
       "P12_root_rep1                                                        NaN\n",
       "P12_root_rep2                                                        NaN\n",
       "P12_root_rep3                                                        NaN\n",
       "P12_stem_rep1                                                        NaN\n",
       "P12_stem_rep2                                                        NaN\n",
       "P12_stem_rep3                                                        NaN\n",
       "P56_leaf_rep1                                                        NaN\n",
       "P56_leaf_rep2                                                        NaN\n",
       "P56_leaf_rep3                                                        NaN\n",
       "P56_medfr_rep1                                                       NaN\n",
       "P56_medfr_rep2                                                       NaN\n",
       "P56_medfr_rep3                                                       NaN\n",
       "P56_oldfr_rep1                                                       NaN\n",
       "P56_oldfr_rep2                                                       NaN\n",
       "P56_oldfr_rep3                                                       NaN\n",
       "P56_root_rep1                                                        NaN\n",
       "P56_root_rep2                                                        NaN\n",
       "P56_root_rep3                                                        NaN\n",
       "P56_stem_rep1                                                        NaN\n",
       "P56_stem_rep2                                                        NaN\n",
       "P56_stem_rep3                                                        NaN\n",
       "P56_youfr_rep1                                                       NaN\n",
       "P56_youfr_rep2                                                       NaN\n",
       "P56_youfr_rep3                                                       NaN\n",
       "Name: Pfim_g20283_i0.p1, dtype: object"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# (~data['pfam'].str.contains('PF00195.23'))\n",
    "\n",
    "data.loc['Pfim_g20283_i0.p1']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Usage: python blastp.py -db <path_to_blast_database> -q <path_to_query_sequence> --mcov 70 --msim 50 --proteome <path_to_proteome> -o <output_fasta>\n",
    "#   -db, --blastDB: Path to the BLAST database.\n",
    "#   -q, --query: Path to the query sequence (FASTA file).\n",
    "#   --proteome: Path to the proteome sequence database (FASTA file).\n",
    "#   --mcov: Minimum coverage for a BLAST hit to be retained (default: 0).\n",
    "#   --msim: Minimum similarity for a BLAST hit to be retained (default: 0).\n",
    "#   -o, --output: Output folder where matching sequences will be stored in a FASTA file.\n",
    "\n",
    "def main():\n",
    "    parser = argparse.ArgumentParser(description=\"XXX\")\n",
    "    parser.add_argument(\"-db\", \"--blastDB\", dest=\"blastDB\", required=True, help=\"Path to BLAST database.\")    \n",
    "    parser.add_argument(\"-q\", \"--query\", dest=\"query\", required=True, help=\"Path to query serquence (FASTA file).\")\n",
    "    parser.add_argument(\"--proteome\", dest=\"proteome\", required=True, help=\"Path to proteome (FASTA file).\")\n",
    "    parser.add_argument(\"--mcov\", dest=\"min_cov\", default=0, help=\"Minimum coverage for a BLAST hit to be retained.\")\n",
    "    parser.add_argument(\"--msim\", dest=\"min_sim\", default=0, help=\"Minimum similarity for a BLAST hit to be retained.\")\n",
    "    parser.add_argument(\"-o\", \"--output\", dest=\"output_fasta\", required=True, help=\"Output folder\")\n",
    "    args = parser.parse_args()\n",
    "\n",
    "    #run blastp\n",
    "    blast_results = blastp(args.blastDB, args.query, min_cov = float(args.min_cov), min_sim = float(args.min_sim))\n",
    "    print(blast_results)\n",
    "    hit_ids = set(blast_results['seq_id'])\n",
    "\n",
    "    #empty list to store the matching sequences\n",
    "    hit_sequences = []\n",
    "\n",
    "    #retreive hit sequences from proteome.pep\n",
    "    with open(args.proteome, 'r') as proteome_handle:\n",
    "        for record in SeqIO.parse(proteome_handle, 'fasta'):\n",
    "            if record.id in hit_ids:\n",
    "                hit_sequences.append(record)\n",
    "\n",
    "    # Write the matching sequences to the output FASTA file\n",
    "    with open(args.output_fasta, 'w') as output_handle:\n",
    "        for record in hit_sequences:\n",
    "            full_id = blast_results.loc[blast_results['seq_id'] == record.id, 'full_id'].values[0]\n",
    "            print(full_id)\n",
    "            record.id = full_id\n",
    "            record.description = \"\"  #clear record description field\n",
    "            SeqIO.write(record, output_handle, 'fasta-2line')\n",
    "            \n",
    "\n",
    "    print(f\"{len(hit_sequences)} BLAST hits stored to '{args.output_fasta}'.\")\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    main()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pipernet",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
