{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "#set cwd to /PiperNET\n",
    "import os\n",
    "os.chdir('..')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from src.rnaseq_utils import get_config\n",
    "import yaml"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "########## Load config.yaml (NEW) ##########\n",
    "def get_config(config_path, **kwargs):\n",
    "    \n",
    "    #load config.yaml as dict\n",
    "    with open(config_path, \"r\") as handle:\n",
    "        config = yaml.safe_load(handle)\n",
    "    \n",
    "    #keep portion of config.yaml based on provided keys \n",
    "    try:\n",
    "        for level in kwargs.values():\n",
    "            config = config[level]\n",
    "        \n",
    "        #convert str paths to Path objects\n",
    "        config = {k: Path(v) if isinstance(v, str) else v for k, v in config.items()}\n",
    "\n",
    "        return config\n",
    "    except KeyError:\n",
    "        raise KeyError(\"Key(s) not found in config.yaml\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'piper09': PosixPath('data/rna-seq/assemblies/piper09/transcriptome.pep'),\n",
       " 'piper10': PosixPath('data/rna-seq/assemblies/piper10/transcriptome.pep'),\n",
       " 'piper12': PosixPath('data/rna-seq/assemblies/piper12/transcriptome.pep'),\n",
       " 'piper20': PosixPath('data/rna-seq/assemblies/piper20/transcriptome.pep'),\n",
       " 'piper23': PosixPath('data/rna-seq/assemblies/piper23/transcriptome.pep')}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load sample folder paths from config.yaml\n",
    "config_path = Path('config/rnaseq_dataprep.yaml')\n",
    "config = get_config(config_path, data='rna-seq', file='assembly')\n",
    "config"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "data/rna-seq/assemblies/piper09/transcriptome.pep\n",
      "data/rna-seq/assemblies/piper09/transcriptome_clstr.pep\n",
      "running CD-HIT on data/rna-seq/assemblies/piper09/transcriptome.pep with identity threshold: 0.98\n",
      "================================================================\n",
      "Program: CD-HIT, V4.8.1, Jan 21 2024, 14:55:34\n",
      "Command: cd-hit -i\n",
      "         data/rna-seq/assemblies/piper09/transcriptome.pep -o\n",
      "         data/rna-seq/assemblies/piper09/transcriptome_clstr.pep\n",
      "         -c 0.98 -n 5 -g 1 a\n",
      "\n",
      "Started: Tue Feb 13 04:17:25 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "\t\t====== CD-HIT version 4.8.1 (built on Jan 21 2024) ======\n",
      "\n",
      "Usage: cd-hit [Options] \n",
      "\n",
      "Options\n",
      "\n",
      "   -i\tinput filename in fasta format, required, can be in .gz format\n",
      "   -o\toutput filename, required\n",
      "   -c\tsequence identity threshold, default 0.9\n",
      " \tthis is the default cd-hit's \"global sequence identity\" calculated as:\n",
      " \tnumber of identical amino acids or bases in alignment\n",
      " \tdivided by the full length of the shorter sequence\n",
      "   -G\tuse global sequence identity, default 1\n",
      " \tif set todata/rna-seq/assemblies/piper10/transcriptome.pep\n",
      "data/rna-seq/assemblies/piper10/transcriptome_clstr.pep\n",
      "running CD-HIT on data/rna-seq/assemblies/piper10/transcriptome.pep with identity threshold: 0.98\n",
      " 0, then use local sequence identity, calculated as :\n",
      " \tnumber of identical amino acids or bases in alignment\n",
      " \tdivided by the length of the alignment\n",
      " \tNOTE!!! don't use -G 0 unless you use alignment coverage controls\n",
      " \tsee options -aL, -AL, -aS, -AS\n",
      "   -b\tband_width of alignment, default 20\n",
      "   -M\tmemory limit (in MB) for the program, default 800; 0 for unlimitted;\n",
      "   -T\tnumber of threads, default 1; with 0, all CPUs will be used\n",
      "   -n\tword_length, default 5, see user's guide for choosing it\n",
      "   -l\tlength of throw_away_sequences, default 10\n",
      "   -t\ttolerance for redundance, default 2\n",
      "   -d\tlength of description in .clstr file, default 20\n",
      " \tif set to 0, it takes the fasta defline and stops at first space\n",
      "   -s\tlength difference cutoff, default 0.0\n",
      " \tif set to 0.9, the shorter sequences need to be\n",
      " \tat least 90% length of the representative of the cluster\n",
      "   -S\tlength difference cutoff in amino acid, default 999999\n",
      " \tif set to 60, the length difference between the shorter sequences\n",
      " \tand the representative of the cluster can not be bigger than 60\n",
      "   -aL\talignment coverage for the longer sequence, default 0.0\n",
      " \tif set to 0.9, the alignment must covers 90% of the sequence\n",
      "   -AL\talignment coverage control for the longer sequence, default 99999999\n",
      " \tif set to 60, and the length of the sequence is 400,\n",
      " \tthen the alignment must be >= 340 (400-60) residues\n",
      "   -aS\talignment coverage for the shorter sequence, default 0.0\n",
      " \tif set to 0.9, the alignment must covers 90% of the sequence\n",
      "   -AS\talignment coverage control for the shorter sequence, default 99999999\n",
      " \tif set to 60, and the length of the sequence is 400,\n",
      " \tthen the alignment must be >= 340 (400-60) residues\n",
      "   -A\tminimal alignment coverage control for the both sequences, default 0\n",
      " \talignment must cover >= this value for both sequences \n",
      "   -uL\tmaximum unmatched percentage for the longer sequence, default 1.0\n",
      " \tif set to 0.1, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10% of the sequence\n",
      "   -uS\tmaximum unmatched percentage for the shorter sequence, default 1.0\n",
      " \tif set to 0.1, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10% of the sequence\n",
      "   -U\tmaximum unmatched length, default 99999999\n",
      " \tif set to 10, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10 bases\n",
      "   -B\t1 or 0, default 0, by default, sequences are stored in RAM\n",
      " \tif set to 1, sequence are stored on hard drive\n",
      " \t!! No longer supported !!\n",
      "   -p\t1 or 0, default 0\n",
      " \tif set to 1, print alignment overlap in .clstr file\n",
      "   -g\t1 or 0, default 0\n",
      " \tby cd-hit's default algorithm, a sequence is clustered to the first \n",
      " \tcluster that meet the threshold (fast cluster). If set to 1, the program\n",
      " \twill cluster it into the most similar cluster that meet the threshold\n",
      " \t(accurate but slow mode)\n",
      " \tbut either 1 or 0 won't change the representatives of final clusters\n",
      "   -sc\tsort clusters by size (number of sequences), default 0, output clusters by decreasing length\n",
      " \tif set to 1, output clusters by decreasing size\n",
      "   -sf\tsort fasta/fastq by cluster size (number of sequences), default 0, no sorting\n",
      " \tif set to 1, output sequences by decreasing cluster size\n",
      " \tthis can be very slow if the input is in .gz format\n",
      "   -bak\twrite backup cluster file (1 or 0, default 0)\n",
      "   -h\tprint this help\n",
      "\n",
      "   Questions, bugs, contact Weizhong Li at liwz@sdsc.edu\n",
      "   For updated versions and information, please visit: http://cd-hit.org\n",
      "                                                    or https://github.com/weizhongli/cdhit\n",
      "\n",
      "   cd-hit web server is also available from http://cd-hit.org\n",
      "\n",
      "   If you find cd-hit useful, please kindly cite:\n",
      "\n",
      "   \"CD-HIT: a fast program for clustering and comparing large sets of protein or nucleotide sequences\", Weizhong Li & Adam Godzik. Bioinformatics, (2006) 22:1658-1659\n",
      "   \"CD-HIT: accelerated for clustering the next generation sequencing data\", Limin Fu, Beifang Niu, Zhengwei Zhu, Sitao Wu & Weizhong Li. Bioinformatics, (2012) 28:3150-3152\n",
      "\n",
      "\n",
      "================================================================\n",
      "Program: CD-HIT, V4.8.1, Jan 21 2024, 14:55:34\n",
      "Command: cd-hit -i\n",
      "         data/rna-seq/assemblies/piper10/transcriptome.pep -o\n",
      "         data/rna-seq/assemblies/piper10/transcriptome_clstr.pep\n",
      "         -c 0.98 -n 5 -g 1 a\n",
      "\n",
      "Started: Tue Feb 13 04:17:25 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "\t\t====== CD-HIT version 4.8.1 (built on Jan 21 2024) ======\n",
      "\n",
      "Usage: cd-hit [Options] \n",
      "\n",
      "Options\n",
      "\n",
      "   -i\tinput filename in fasta format, required, can be in .gz format\n",
      "   -o\toutput filename, required\n",
      "   -c\tsequence identity threshold, default 0.9\n",
      " \tthis is the default cd-hit's \"global sequence identity\" calculated as:\n",
      " \tnumber of identical amino acids or bases in alignment\n",
      " \tdivided by the full length of the shorter sequence\n",
      "   -G\tuse global sequence identity, default 1\n",
      " \tif set to 0, then use local sequence identity, calculated as :\n",
      " \tnumber of identical amino acids or bases in alignment\n",
      " \tdivided by the length of the alignment\n",
      " \tNOTE!!! don't use -G 0 unless you use alignment coverage controls\n",
      " \tsee options -aL, -AL, -aS, -AS\n",
      "   -b\tband_width of alignment, default 20\n",
      "   -M\tmemory limit (in MB) for the program, default 800; 0 for unlimitted;\n",
      "   -T\tnumber of threads, default 1; with 0, all CPUs will be used\n",
      "   -n\tword_length, default 5, see user's guide for choosing it\n",
      "   -l\tlength of throw_away_sequences, default 10\n",
      "   -t\ttolerance for redundance, default 2\n",
      "   -d\tlength of description in .clstr file, default 20\n",
      " \tif set to 0, it takes the fasta defline and stops at first space\n",
      "   -s\tlength difference cutoff, default 0.0\n",
      " \tif set to 0.9, the shorter sequences need to be\n",
      " \tat least 90% length of the representative of the cluster\n",
      "   -S\tlength difference cutoff in amino acid, default 999999\n",
      " \tif set to 60, the length difference between the shorter sequences\n",
      " \tand the representative of the cluster can not be bigger than 60\n",
      "   -aL\talignment coverage for the longer sequence, default 0.0\n",
      " \tif set to 0.9, the alignment must covers 90% of the sequence\n",
      "   -AL\talignment coverage control for the longer sequence, default 99999999\n",
      " \tif set to 60, and the length of the sequence is 400,\n",
      " \tthen the alignment must be >= 340 (400-60) residues\n",
      "   -aS\talignment coverage for the shorter sequence, default 0.0\n",
      " \tif set to 0.9, the alignment must covers 90% of the sequence\n",
      "   -AS\talignment coverage control for the shorter sequence, default 99999999\n",
      " \tif set to 60, and the length of the sequence is 400,\n",
      " \tthen the alignment must be >= 340 (400-60) residues\n",
      "   -A\tminimal alignment coverage control for the both sequences, default 0\n",
      " \talignment must cover >= this value for both sequences \n",
      "   -uL\tmaximum unmatched percentage for the longer sequence, default 1.0\n",
      " \tif set to 0.1, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10% of the sequence\n",
      "   -uS\tmaximum unmatched percentage for the shorter sequence, default 1.0\n",
      " \tif set to 0.1, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10% of the sequence\n",
      "   -U\tmaximum unmatched length, default 99999999\n",
      " \tif set to 10, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10 bases\n",
      "   -B\t1 or 0, default 0, by default, sequences are stored in RAM\n",
      " \tif set to 1, sequence are stored on hard drive\n",
      " \t!! No longer supported !!\n",
      "   -p\t1 or 0, default 0\n",
      " \tif set to 1, print alignment overlap in .clstr file\n",
      "   -g\t1 or 0, default 0\n",
      " \tby cd-hit's default algorithm, a sequence is clustered to the first \n",
      " \tcluster that meet the threshold (fast cluster). If set to 1, the program\n",
      " \twill cluster it into the most similar cluster that meet the threshold\n",
      " \t(accurate but slow mode)\n",
      " \tbut either 1 or 0 won't change the representatives of final clusters\n",
      "   -sc\tsort clusters by size (number of sequences), default 0, output clusters by decreasing length\n",
      " \tif set to 1, output clusters by decreasing size\n",
      "   -sf\tsort fasta/fastq by cluster size (number of sequences), default 0, no sorting\n",
      " \tif set to 1, output sequences by decreasing cluster size\n",
      " \tthis can be very slow if the input is in .gz format\n",
      "   -bak\twrite backup cluster file (1 or 0, default 0)\n",
      "   -h\tprint this help\n",
      "\n",
      "   Questions, bugs, contact Weizhong Li at liwz@sdsc.edu\n",
      "   For updated versions and information, please visit: http://cd-hit.org\n",
      "                                                    or https://github.com/weizhongli/cdhit\n",
      "\n",
      "   cd-hit web server is also available from http://cd-hit.org\n",
      "\n",
      "   If you find cd-hit useful, please kindly cite:\n",
      "\n",
      "   \"CD-HIT: a fast program for clustering and comparing large sets of protein or nucleotide sequences\", Weizhong Li & Adam Godzik. Bioinformatics, (2006) 22:1658-1659\n",
      "   \"CD-HIT: accelerated for clustering the next generation sequencing data\", Limin Fu, Beifang Niu, Zhengwei Zhu, Sitao Wu & Weizhong Li. Bioinformatics, (2012) 28:3150-3152\n",
      "\n",
      "\n",
      "data/rna-seq/assemblies/piper12/transcriptome.pep\n",
      "data/rna-seq/assemblies/piper12/transcriptome_clstr.pep\n",
      "running CD-HIT on data/rna-seq/assemblies/piper12/transcriptome.pep with identity threshold: 0.98\n",
      "================================================================\n",
      "Program: CD-HIT, V4.8.1, Jan 21 2024, 14:55:34\n",
      "Command: cd-hit -i\n",
      "         data/rna-seq/assemblies/piper12/transcriptome.pep -o\n",
      "         data/rna-seq/assemblies/piper12/transcriptome_clstr.pep\n",
      "         -c 0.98 -n 5 -g 1 a\n",
      "\n",
      "Started: Tue Feb 13 04:17:25 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "\t\t====== CD-HIT version 4.8.1 (built on Jan 21 2024) ======\n",
      "\n",
      "Usage: cd-hit [Options] \n",
      "\n",
      "Options\n",
      "\n",
      "   -i\tinput filename in fasta format, required, can be in .gz format\n",
      "   -o\toutput filename, required\n",
      "   -c\tsequence identity threshold, default 0.9\n",
      " \tthis is the default cd-hit's \"global sequence identity\" calculated as:\n",
      " \tnumber of identical amino acids or bases in alignment\n",
      " \tdivided by the full length of the shorter sequence\n",
      "   -G\tuse global sequence identity, default 1\n",
      " \tif set to 0, then use local sequence identity, calculated as :\n",
      " \tnumber of identical amino acids or bases in alignment\n",
      " \tdivided by the length of the alignment\n",
      " \tNOTE!!! don't use -G 0 unless you use alignment coverage controls\n",
      " \tsee options -aL, -AL, -aS, -AS\n",
      "   -b\tband_width of alignment, default 20\n",
      "   -M\tmemory limit (in MB) for the program, default 800; 0 for unlimitted;\n",
      "   -T\tnumber of threads, default 1; with 0, all CPUs will be used\n",
      "   -n\tword_length, default 5, see user's guide for choosing it\n",
      "   -l\tlength of throw_away_sequences, default 10\n",
      "   -t\ttolerance for redundance, default 2\n",
      "   -d\tlength of description in .clstr file, default 20\n",
      " \tif set to 0, it takes the fasta defline and stops at first space\n",
      "   -s\tlength difference cutoff, default 0.0\n",
      " \tif set to 0.9, the shorter sequences need to be\n",
      " \tat least 90% length of the representative of the cluster\n",
      "   -S\tlength difference cutoff in amino acid, default 999999\n",
      " \tif set to 60, the length difference between the shorter sequences\n",
      " \tand the representative of the cluster can not be bigger than 60\n",
      "   -aL\talignment coverage for the longer sequence, default 0.0\n",
      " \tif set to 0.9, the alignment must covers 90% of the sequence\n",
      "   -AL\talignment coverage control for the longer sequence, default 99999999\n",
      " \tif set to 60, and the length of the sequence is 400,\n",
      " \tthen the alignment must be >= 340 (400-60) residues\n",
      "   -aS\talignment coverage for the shorter sequence, default 0.0\n",
      " \tif set to 0.9, the alignment must covers 90% of the sequence\n",
      "   -AS\talignment coverage control for the shorter sequence, default 99999999\n",
      " \tif set to 60, and the length of the sequence is 400,\n",
      " \tthen the alignment must be >= 340 (400-60) residues\n",
      "   -A\tminimal alignment coverage control for the both sequences, default 0\n",
      " \talignment must cover >= this value for both sequences \n",
      "   -uL\tmaximum unmatched percentage for the longer sequence, default 1.0\n",
      " \tif set to 0.1, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10% of the sequence\n",
      "   -uS\tmaximum unmatched percentage for the shorter sequence, default 1.0\n",
      " \tif set to 0.1, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10% of the sequence\n",
      "   -U\tmaximum unmatched length, default 99999999\n",
      " \tif set to 10, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10 bases\n",
      "   -B\t1 or 0, default 0, by default, sequences are stored in RAM\n",
      " \tif set to 1, sequence are stored on hard drive\n",
      " \t!! No longer supported !!\n",
      "   -p\t1 or 0, default 0\n",
      " \tif set to 1, print alignment overlap in .clstr file\n",
      "   -g\t1 or 0, default 0\n",
      " \tby cd-hit's default algorithm, a sequence is clustered to the first \n",
      " \tcluster that meet the threshold (fast cluster). If set to 1, the program\n",
      " \twill cluster it into the most similar cluster that meet the threshold\n",
      " \t(accurate but slow mode)\n",
      " \tbut either 1 or 0 won't change the representatives of final clusters\n",
      "   -sc\tsort clusters by size (number of sequences), default 0, output clusters by decreasing length\n",
      " \tif set to 1, output clusters by decreasing size\n",
      "   -sf\tsort fasta/fastq by cluster size (number of sequences), default 0, no sorting\n",
      " \tif set to 1, output sequences by decreasing cluster size\n",
      " \tthis can be very slow if the input is in .gz format\n",
      "   -bak\twrite backup cluster file (1 or 0, default 0)\n",
      "   -h\tprint this help\n",
      "\n",
      "   Questions, bugs, contact Weizhong Li at liwz@sdsc.edu\n",
      "   For updated versions and information, please visit: http://cd-hit.org\n",
      "                                                    or https://github.com/weizhongli/cdhit\n",
      "\n",
      "   cd-hit web server is also available from http://cd-hit.org\n",
      "\n",
      "   If you find cd-hit useful, please kindly cite:\n",
      "\n",
      "   \"CD-HIT: a fast program for clustering and comparing large sets of protein or nucleotide sequences\", Weizhong Li & Adam Godzik. Bioinformatics, (2006) 22:1658-1659\n",
      "   \"CD-HIT: accelerated for clustering the next generation sequencing data\", Limin Fu, Beifang Niu, Zhengwei Zhu, Sitao Wu & Weizhong Li. Bioinformatics, (2data/rna-seq/assemblies/piper20/transcriptome.pep\n",
      "data/rna-seq/assemblies/piper20/transcriptome_clstr.pep\n",
      "running CD-HIT on data/rna-seq/assemblies/piper20/transcriptome.pep with identity threshold: 0.98\n",
      "012) 28:3150-3152\n",
      "\n",
      "\n",
      "================================================================\n",
      "Program: CD-HIT, V4.8.1, Jan 21 2024, 14:55:34\n",
      "Command: cd-hit -i\n",
      "         data/rna-seq/assemblies/piper20/transcriptome.pep -o\n",
      "         data/rna-seq/assemblies/piper20/transcriptome_clstr.pep\n",
      "         -c 0.98 -n 5 -g 1 a\n",
      "\n",
      "Started: Tue Feb 13 04:17:25 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "\t\t====== CD-HIT version 4.8.1 (built on Jan 21 2024) ======\n",
      "\n",
      "Usage: cd-hit [Options] \n",
      "\n",
      "Options\n",
      "\n",
      "   -i\tinput filename in fasta format, required, can be in .gz format\n",
      "   -o\toutput filename, required\n",
      "   -c\tsequence identity threshold, default 0.9\n",
      " \tthis is the default cd-hit's \"global sequence identity\" calculated as:\n",
      " \tnumber of identical amino acids or bases in alignment\n",
      " \tdivided by the full length of the shorter sequence\n",
      "   -G\tuse global sequence identity, default 1\n",
      " \tif set to 0, then use local sequence identity, calculated as :\n",
      " \tnumber of identical amino acids or bases in alignment\n",
      " \tdivided by the length of the alignment\n",
      " \tNOTE!!! don't use -G 0 unless you use alignment coverage controls\n",
      " \tsee options -aL, -AL, -aS, -AS\n",
      "   -b\tband_width of alignment, default 20\n",
      "   -M\tmemory limit (in MB) for the program, default 800; 0 for unlimitted;\n",
      "   -T\tnumber of threads, default 1; with 0, all CPUs will be used\n",
      "   -n\tword_length, default 5, see user's guide for choosing it\n",
      "   -l\tlength of throw_away_sequences, default 10\n",
      "   -t\ttolerance for redundance, default 2\n",
      "   -d\tlength of description in .clstr file, default 20\n",
      " \tif set to 0, it takes the fasta defline and stops at first space\n",
      "   -s\tlength difference cutoff, default 0.0\n",
      " \tif set to 0.9, the shorter sequences need to be\n",
      " \tat least 90% length of the representative of the cluster\n",
      "   -S\tlength difference cutoff in amino acid, default 999999\n",
      " \tif set to 60, the length difference between the shorter sequences\n",
      " \tand the representative of the cluster can not be bigger than 60\n",
      "   -aL\talignment coverage for the longer sequence, default 0.0\n",
      " \tif set to 0.9, the alignment must covers 90% of the sequence\n",
      "   -AL\talignment coverage control for the longer sequence, default 99999999\n",
      " \tif set to 60, and the length of the sequence is 400,\n",
      " \tthen the alignment must be >= 340 (400-60) residues\n",
      "   -aS\talignment coverage for the shorter sequence, default 0.0\n",
      " \tif set to 0.9, the alignment must covers 90% of the sequence\n",
      "   -AS\talignment coverage control for the shorter sequence, default 99999999\n",
      " \tif set to 60, and the length of the sequence is 400,\n",
      " \tthen the alignment must be >= 340 (400-60) residues\n",
      "   -A\tminimal alignment coverage control for the both sequences, default 0\n",
      " \talignment must cover >= this value for both sequences \n",
      "   -uL\tmaximum unmatched percentage for the longer sequence, default 1.0\n",
      " \tif set to 0.1, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10% of the sequence\n",
      "   -uS\tmaximum unmatched percentage for the shorter sequence, default 1.0\n",
      " \tif set to 0.1, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10% of the sequence\n",
      "   -U\tmaximum unmatched length, default 99999999\n",
      " \tif set to 10, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10 bases\n",
      "   -B\t1 or 0, default 0, by default, sequences are stored in RAM\n",
      " \tif set to 1, sequence are stored on hard drive\n",
      " \t!! No longer supported !!\n",
      "   -p\t1 or 0, default 0\n",
      " \tif set to 1, print alignment overlap in .clstr file\n",
      "   -g\t1 or 0, default 0\n",
      " \tby cd-hit's default algorithm, a sequence is clustered to the first \n",
      " \tcluster that meet the threshold (fast cluster). If set to 1, the program\n",
      " \twill cluster it into the most similar cluster that meet the threshold\n",
      " \t(accurate but slow mode)\n",
      " \tbut either 1 or 0 won't change the representatives of final clusters\n",
      "   -sc\tsort clusters by size (number of sequences), default 0, output clusters by decreasing length\n",
      " \tif set to 1, output clusters by decreasing size\n",
      "   -sf\tsort fasta/fastq by cluster size (number of sequences), default 0, no sorting\n",
      " \tif set to 1, output sequences by decreasing cluster size\n",
      " \tthis can be very slow if the input is in .gz format\n",
      "   -bak\twrite backup cluster file (1 or 0, default 0)\n",
      "   -h\tprint this help\n",
      "\n",
      "   Questions, bugs, contact Weizhong Li at liwz@sdsc.edu\n",
      "   For updated versions and information, please visit: http://cd-hit.org\n",
      "                                                    or https://github.com/weizhongli/cdhit\n",
      "\n",
      "   cd-hit web server is also available from http://cd-hit.org\n",
      "\n",
      "   If you find cd-hit useful, please kindly cite:\n",
      "\n",
      "   \"CD-HIT: a fast program for clustering and comparing large sets of protein or nucleotide sequences\", Weizhong Li & Adam Godzik. Bioinformatics, (2006) 22:1658-1659\n",
      "   \"CD-HIT: accelerated for clustering the next generation sequencing data\", Limin Fu, Beifang Niu, Zhengwei Zhu, Sitao Wu & Weizhong Li. Bioinformatics, (2012) 28:3150-3152\n",
      "\n",
      "\n",
      "data/rna-seq/assemblies/piper23/transcriptome.pep\n",
      "data/rna-seq/assemblies/piper23/transcriptome_clstr.pep\n",
      "running CD-HIT on data/rna-seq/assemblies/piper23/transcriptome.pep with identity threshold: 0.98\n",
      "================================================================\n",
      "Program: CD-HIT, V4.8.1, Jan 21 2024, 14:55:34\n",
      "Command: cd-hit -i\n",
      "         data/rna-seq/assemblies/piper23/transcriptome.pep -o\n",
      "         data/rna-seq/assemblies/piper23/transcriptome_clstr.pep\n",
      "         -c 0.98 -n 5 -g 1 a\n",
      "\n",
      "Started: Tue Feb 13 04:17:25 2024\n",
      "================================================================\n",
      "                            Output                              \n",
      "----------------------------------------------------------------\n",
      "\t\t====== CD-HIT version 4.8.1 (built on Jan 21 2024) ======\n",
      "\n",
      "Usage: cd-hit [Options] \n",
      "\n",
      "Options\n",
      "\n",
      "   -i\tinput filename in fasta format, required, can be in .gz format\n",
      "   -o\toutput filename, required\n",
      "   -c\tsequence identity threshold, default 0.9\n",
      " \tthis is the default cd-hit's \"global sequence identity\" calculated as:\n",
      " \tnumber of identical amino acids or bases in alignment\n",
      " \tdivided by the full length of the shorter sequence\n",
      "   -G\tuse global sequence identity, default 1\n",
      " \tif set to 0, then use local sequence identity, calculated as :\n",
      " \tnumber of identical amino acids or bases in alignment\n",
      " \tdivided by the length of the alignment\n",
      " \tNOTE!!! don't use -G 0 unless you use alignment coverage controls\n",
      " \tsee options -aL, -AL, -aS, -AS\n",
      "   -b\tband_width of alignment, default 20\n",
      "   -M\tmemory limit (in MB) for the program, default 800; 0 for unlimitted;\n",
      "   -T\tnumber of threads, default 1; with 0, all CPUs will be used\n",
      "   -n\tword_length, default 5, see user's guide for choosing it\n",
      "   -l\tlength of throw_away_sequences, default 10\n",
      "   -t\ttolerance for redundance, default 2\n",
      "   -d\tlength of description in .clstr file, default 20\n",
      " \tif set to 0, it takes the fasta defline and stops at first space\n",
      "   -s\tlength difference cutoff, default 0.0\n",
      " \tif set to 0.9, the shorter sequences need to be\n",
      " \tat least 90% length of the representative of the cluster\n",
      "   -S\tlength difference cutoff in amino acid, default 999999\n",
      " \tif set to 60, the length difference between the shorter sequences\n",
      " \tand the representative of the cluster can not be bigger than 60\n",
      "   -aL\talignment coverage for the longer sequence, default 0.0\n",
      " \tif set to 0.9, the alignment must covers 90% of the sequence\n",
      "   -AL\talignment coverage control for the longer sequence, default 99999999\n",
      " \tif set to 60, and the length of the sequence is 400,\n",
      " \tthen the alignment must be >= 340 (400-60) residues\n",
      "   -aS\talignment coverage for the shorter sequence, default 0.0\n",
      " \tif set to 0.9, the alignment must covers 90% of the sequence\n",
      "   -AS\talignment coverage control for the shorter sequence, default 99999999\n",
      " \tif set to 60, and the length of the sequence is 400,\n",
      " \tthen the alignment must be >= 340 (400-60) residues\n",
      "   -A\tminimal alignment coverage control for the both sequences, default 0\n",
      " \talignment must cover >= this value for both sequences \n",
      "   -uL\tmaximum unmatched percentage for the longer sequence, default 1.0\n",
      " \tif set to 0.1, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10% of the sequence\n",
      "   -uS\tmaximum unmatched percentage for the shorter sequence, default 1.0\n",
      " \tif set to 0.1, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10% of the sequence\n",
      "   -U\tmaximum unmatched length, default 99999999\n",
      " \tif set to 10, the unmatched region (excluding leading and tailing gaps)\n",
      " \tmust not be more than 10 bases\n",
      "   -B\t1 or 0, default 0, by default, sequences are stored in RAM\n",
      " \tif set to 1, sequence are stored on hard drive\n",
      " \t!! No longer supported !!\n",
      "   -p\t1 or 0, default 0\n",
      " \tif set to 1, print alignment overlap in .clstr file\n",
      "   -g\t1 or 0, default 0\n",
      " \tby cd-hit's default algorithm, a sequence is clustered to the first \n",
      " \tcluster that meet the threshold (fast cluster). If set to 1, the program\n",
      " \twill cluster it into the most similar cluster that meet the threshold\n",
      " \t(accurate but slow mode)\n",
      " \tbut either 1 or 0 won't change the representatives of final clusters\n",
      "   -sc\tsort clusters by size (number of sequences), default 0, output clusters by decreasing length\n",
      " \tif set to 1, output clusters by decreasing size\n",
      "   -sf\tsort fasta/fastq by cluster size (number of sequences), default 0, no sorting\n",
      " \tif set to 1, output sequences by decreasing cluster size\n",
      " \tthis can be very slow if the input is in .gz format\n",
      "   -bak\twrite backup cluster file (1 or 0, default 0)\n",
      "   -h\tprint this help\n",
      "\n",
      "   Questions, bugs, contact Weizhong Li at liwz@sdsc.edu\n",
      "   For updated versions and information, please visit: http://cd-hit.org\n",
      "                                                    or https://github.com/weizhongli/cdhit\n",
      "\n",
      "   cd-hit web server is also available from http://cd-hit.org\n",
      "\n",
      "   If you find cd-hit useful, please kindly cite:\n",
      "\n",
      "   \"CD-HIT: a fast program for clustering and comparing large sets of protein or nucleotide sequences\", Weizhong Li & Adam Godzik. Bioinformatics, (2006) 22:1658-1659\n",
      "   \"CD-HIT: accelerated for clustering the next generation sequencing data\", Limin Fu, Beifang Niu, Zhengwei Zhu, Sitao Wu & Weizhong Li. Bioinformatics, (2012) 28:3150-3152\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "\n",
    "import subprocess\n",
    "import argparse\n",
    "from src.rnaseq_utils import get_config\n",
    "from pathlib import Path\n",
    "\n",
    "\n",
    "config_path = Path('config/rnaseq_dataprep.yaml')\n",
    "\n",
    "#load sample folder paths from config.yaml\n",
    "assembly_paths = get_config(config_path, data='rna-seq', file='assembly')\n",
    "\n",
    "#iterate over path dictionary\n",
    "for input in assembly_paths.values():\n",
    "    print(input)\n",
    "    \n",
    "    try:\n",
    "        #i/o file paths\n",
    "        output = Path(input).with_name('transcriptome_clstr.pep')\n",
    "        print(output)\n",
    "\n",
    "        #construct command\n",
    "        min_identity = str(0.98)\n",
    "        print(f'running CD-HIT on {input} with identity threshold: {min_identity}')\n",
    "        command = f\"cd-hit -i {input} -o {output} -c {min_identity} -n 5 -g 1 a\"\n",
    "\n",
    "        #run command\n",
    "        subprocess.run(command, shell=True)\n",
    "\n",
    "    #skip to next file if input is not found\n",
    "    except FileNotFoundError:\n",
    "        print(f'{input} not found. Skipping to next sample.')\n",
    "        continue"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data/rna-seq/assemblies/piper09/transcriptome.pep\n",
    "data/rna-seq/assemblies/piper09/transcriptome.pep"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pipernet",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
